design/lifecycle.md

```diff
@@ -0,0 +1,195 @@
+> Note! This could be improved by doing these intersections for each rule independently, It could potentially reduce the number of commits, I believe it is not worth the complication for now.   
+
+lakeFS will receive this data and return/write all the potential commits (with commit branches) for this run.
+
+#### Find addresses that should be removed by lifecycle policy
+
+From the previous stage we got a table of the following form commit_id | branches
+Using that we:
+1. get rang_id from commit_id
+2. group by range_id selecting the union of branches
+3. get (key, address) from ranges
+4. group by (keys, address) selecting the union of branches
+5. select all addresses matching the policy rules (which did not match policy rule last run)
+
+#### Find deduplications
+
+##### Definitions
+deduplications - for a given path P with address A, a deduplication would be a different path P` referencing the same address A
+
+In the case of lifecycle, before removing an address, we would first need to check that it has no deduplications that should not be removed as part of lifecycle.
+For example:
+Assuming we have rule for prefix `foo/bar` and for prefix `foo/tar`
+- `foo/bar/a 1`
+- `foo/bar/b 4`
+- `foo/tar/a 1`
+- `foo/tar/b 2`
+- `foo/other/c 2`
+
+In the given case:
+- 4 could be removed because it has no deduplications
+- 1 has deduplications but it could be removed due to the fact that all deduplications should be removed by lifecycle
+- 2 can't be removed because it has a deduplication that should be accessible after lifecycle job is done (foo/other/c)
+
+> **_NOTE:_**  In case an address wasn't removed due to deduplication it will not be removed in the following run even if the deduplication was deleted. i.e regarding the previous example address 2 wasn't removed due to deduplication, in case foo/other/c was somehow deleted, when running the job again, we might assume address 2 should now be removed because it matches lifecycle policy and has no deduplication. We will not support that case
+
+##### Algorithm
+1. For every address check if it is a deduplication:
+      - If it appears in commits that are not in the potential commits -> it's a deduplication
+      - If it appears in a commit that is a `potential commit` in a path that does not match any rule -> it's a deduplication
+2. filter the deduplications out from the addresses to be removed
+3. write the result as a table ( run_id, address, paths, matched_rules) to  `_lakefs/retention/lc/addresses`  partitioned by date
+
+
+#### Delete addresses
+The last step would be to delete all the addresses that should be removed. this step was already implemented as part of GC.
+After deleting we will write some kind of success/status file to `_laefs/retention/lc/stats/[date].csv`
+
+> **_NOTE:_** This algorithm could be optimized when running together with garbage collection by working with only `active commits` (and not all commits, both in finding potential commits and in finding deduplications)
```

Comment:
This means we currently do not handle uncommitted entries anywhere? (which is ok as a starting point, just want to clarify)